### DISCUSSION TITLE
Time-dependent absorption in nlmixr2/RxODE

### DISCUSSION THREAD

**Discussion by Murali-Ramanathan** [nlmixr2/nlmixr2#292] [Q&A] [answered]
I have a one-compartment model with dose-linear but time-dependent input.
`d/dt(centr)  =  Dose*PA*input_function(t, TA) -KA*centr`
where _PA_, _TA_, _KA_ are parameters. 

I was able to simulate the time course following a single bolus dose by simply setting the value of _Dose_ in the input function with _deSolve_. How would I set up a multiple bolus dosing simulation in RxODE?

**Comment by john-harrold** [comment, marked as answer]
The example below will dose into a depot compartment and then you can alter the absorption rate using any arbitrary function of time. Alternatively you could make that a covariate in your dataset and use that. 

``` r
library(rxode2)
#> rxode2 3.0.2 using 1 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
#> ========================================
#> rxode2 has not detected OpenMP support and will run in single-threaded mode
#> This is a Mac. Please read https://mac.r-project.org/openmp/
#> ========================================
mod = function () 
{
    description <- "One compartment PK model with linear clearance using differential equations"
    dosing <- c("central", "depot")
    ini({
        lka <- 0.45
        label("Absorption rate (Ka)")
        lcl <- 1
        label("Clearance (CL)")
        lvc <- 3.45
        label("Central volume of distribution (V)")
        propSd <- c(0, 0.5)
        label("Proportional residual error (fraction)")
    })
    model({
        ka <- exp(lka)
        cl <- exp(lcl)
        vc <- exp(lvc)
        kel <- cl/vc
        # you need to put your time-dependent logic here or use a covariate in
        # your dataset
        ka_t <-ka*(1+exp(-.1*time))
        d/dt(depot) <- -ka_t * depot
        d/dt(central) <- ka_t * depot - kel * central
        Cc <- central/vc
        Cc ~ prop(propSd)
    })
}


ev = et(seq(0,7*10,7), amt=1, ii=7, cmt="depot")
#> Warning: 'ii' requires non zero additional doses ('addl') or steady state
#> dosing ('ii': 7.000000, 'ss': 0; 'addl': 0), reset 'ii' to zero

res = rxSolve(mod, ev, nSub=1)
#> ℹ parameter labels from comments are typically ignored in non-interactive mode
#> ℹ Need to run with the source intact to parse comments
#> using C compiler: ‘Apple clang version 15.0.0 (clang-1500.3.9.4)’
#> using SDK: ‘’

plot(res)
```

![](https://i.imgur.com/KdglsiH.png)

<sup>Created on 2024-11-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by Murali-Ramanathan** [reply]
Thanks, John.

**Reply by Murali-Ramanathan** [reply]
I just noted the decline in the peak heights in `depot`. Is the `time` dependence of each bolus being computed relative to its administration time for the `ka_t` calculations? I think it may be using the time relative to `t = 0` for the `ka_t` from the multiple doses?

**Reply by mattfidler** [reply]
You can use `tad0()` for time relative to last dose (or time after dose; 0 is to use 0 before doses instead of `NA`):


``` r
library(rxode2)
#> rxode2 3.0.2.9000 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
mod = function ()
{
  description <- "One compartment PK model with linear clearance using differential equations"
  dosing <- c("central", "depot")
  ini({
    lka <- 0.45
    label("Absorption rate (Ka)")
    lcl <- 1
    label("Clearance (CL)")
    lvc <- 3.45
    label("Central volume of distribution (V)")
    propSd <- c(0, 0.5)
    label("Proportional residual error (fraction)")
  })
  model({
    ka <- exp(lka)
    cl <- exp(lcl)
    vc <- exp(lvc)
    kel <- cl/vc
    # you need to put your time-dependent logic here or use a covariate in
    # your dataset
    ka_t <-ka*(1+exp(-.1*tad0()))
    d/dt(depot) <- -ka_t * depot
    d/dt(central) <- ka_t * depot - kel * central
    Cc <- central/vc
    Cc ~ prop(propSd)
  })
}


ev = et(0, amt=1, ii=7, until=7*10, cmt="depot")

res = rxSolve(mod, ev, nSub=1)
#> ℹ parameter labels from comments are typically ignored in non-interactive mode
#> ℹ Need to run with the source intact to parse comments

plot(res, depot, central, Cc, sim, ka_t)
```

![](https://i.imgur.com/2dK0iBf.png)

<sup>Created on 2024-11-30 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by john-harrold** [reply]
Without tad0
<img width="185" alt="image" src="https://github.com/user-attachments/assets/4eaf9e15-b86e-440f-8338-46aec9a62dee">

With tad0
<img width="223" alt="image" src="https://github.com/user-attachments/assets/0a79eb16-2a10-4d7a-a81f-73ce2ec762f2">


I think they are referring to this which I believe is just an artifact of automated sampling.

**Reply by mattfidler** [reply]
Hi @john-harrold  you are right that the depot is an artifact of the sample times.

**Comment by Murali-Ramanathan** [comment]
Thanks, Matt!

From: Matthew Fidler ***@***.***>
Date: Saturday, November 30, 2024 at 10:49 AM
To: nlmixr2/nlmixr2 ***@***.***>
Cc: Murali Ramanathan ***@***.***>, Author ***@***.***>
Subject: Re: [nlmixr2/nlmixr2] Time-dependent absorption in nlmixr2/RxODE (Discussion #292)

You can use tad0() for time relative to last dose (or time after dose; 0 is to use 0 before doses instead of NA):

library(rxode2)

#> rxode2 3.0.2.9000 using 8 threads (see ?getRxThreads)

#>   no cache: create with `rxCreateCache()`

mod = function ()

{

  description <- "One compartment PK model with linear clearance using differential equations"

  dosing <- c("central", "depot")

  ini({

    lka <- 0.45

    label("Absorption rate (Ka)")

    lcl <- 1

    label("Clearance (CL)")

    lvc <- 3.45

    label("Central volume of distribution (V)")

    propSd <- c(0, 0.5)

    label("Proportional residual error (fraction)")

  })

  model({

    ka <- exp(lka)

    cl <- exp(lcl)

    vc <- exp(lvc)

    kel <- cl/vc

    # you need to put your time-dependent logic here or use a covariate in

    # your dataset

    ka_t <-ka*(1+exp(-.1*tad0()))

    d/dt(depot) <- -ka_t * depot

    d/dt(central) <- ka_t * depot - kel * central

    Cc <- central/vc

    Cc ~ prop(propSd)

  })

}





ev = et(0, amt=1, ii=7, until=7*10, cmt="depot")



res = rxSolve(mod, ev, nSub=1)

#> ℹ parameter labels from comments are typically ignored in non-interactive mode

#> ℹ Need to run with the source intact to parse comments



plot(res, depot, central, Cc, sim, ka_t)

[Image removed by sender.]<https://camo.githubusercontent.com/89cba686bcb33ffc9e9a2e27561cb98645e71027b57187842900545f0816c8d0/68747470733a2f2f692e696d6775722e636f6d2f32644b306942662e706e67>

Created on 2024-11-30 with reprex v2.1.1<https://reprex.tidyverse.org/>

**Comment by Murali-Ramanathan** [comment]
Dear Matt

I have resolved my install issues. I had to go back and update my R version, go through the installation process. I am back to being able to run.

Is there a way to multi-thread on a MacBook Pro?

Best regards,
Murali

From: Murali Ramanathan ***@***.***>
Date: Wednesday, July 9, 2025 at 1:48 PM
To: nlmixrdevelopment/nlmixr ***@***.***>
Subject: nlmixr2 Install
Dear Matt

I had a working version of nlmixr on my MacBook Pro.

However, my nlmixr2 install stopped working when I tried to reinstall nlmixr2est thinking it would address a multi-threading issue.

I have tried uninstalling and installing but keep running into issues. Here are the error messages.

Error: package or namespace load failed for ‘nlmixr2’:
.onLoad failed in loadNamespace() for 'nlmixr2est', details:
  call: NULL
  error: '.lbfgsb3cPtr' is not an exported object from 'namespace:lbfgsb3c'
In addition: Warning message:
package ‘nlmixr2’ was built under R version 4.3.3

Please help!

Best regards,
Murali

**Reply by mattfidler** [reply]
Hi @Murali-Ramanathan   

You will have to update all of your packages.  In particular you will need to re-install
- `dparser` (probably already installed)
- `rxode2` (probably already installed)
- `nlmixr2est`
- `PreciseSums`
- `lbfgsb3c`
- `nlmixr2`

You may  want to install other packages too.

`nlmixr2CheckInstall()` in theory should help with this.
